core-api/src/main/java/com/optimizely/ab/config/audience/match/AbstractAttributeCondition.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2019, Optimizely and contributors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.optimizely.ab.config.audience.match;
+
+import com.optimizely.ab.config.ProjectConfig;
+import com.optimizely.ab.config.audience.Condition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Base class for {@link Condition} implementations that evaluate an untyped
+ * {@link Object} value associated to an attribute.
+ *
+ * Defines an interface to evaluate match in 3 steps:
+ * <ol>
+ *   <li>extracts the untyped attribute value to evaluate</li>
+ *   <li>filter the attribute to types supported by match logic</li>
+ *   <li>evaluate whether typed input is a match</li>
+ * </ol>
+ *
+ * @param <T> type of value that input is matched against
+ * @param <U> type of converted input that is required to match
+ */
+public abstract class AbstractAttributeCondition<T, U> implements Condition<T> {
```

Comment:
Semantically (and maybe even architecturally 😕) it's not clear to me how this differs from UserAttribute.